### PR TITLE
Added power tracking api

### PR DESCRIPTION
**Pull request by purdeaandrei**
## Description

This has now been tested on chibios, running on an STM32F446, and atmega32u4, and atmega32u2.

## Types of Changes

- [x] Core
- [ ] Bugfix
- [x] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by purdeaandrei** [comment]
Tested on STM32F446, and atmega32u4.

**Review by drashna — CHANGES_REQUESTED** [review]

File: quantum/power.c [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+ */
+
+#include "power.h"
+
+enum power_state power_state;
+
+__attribute__((weak)) void notify_power_state_change_kb(void) {
+}
+
+static void notify_power_state_change(void)
```
[the remark this answers is not in the document]
Comment (purdeaandrei) [reply quantum/power.c]: @drashna 
I made it pass along the state's value, and added a ```_user()``` function.
Could you clarify how you want the quantum function? should I just rename ```notify_power_state_change``` to ```notify_power_state_change_quantum```?
The qmk codebase seems to be inconsistent in this regard.

**Comment by drashna** [comment]
While the code looks good, "power tracking" doesn't sound right to me.  

The point is getting the status from USB, correct?  Then maybe "USB Power State" would be more accurate, maybe? 

That said, I know that naming stuff is hard.  I have some truly atrociously named stuff that has made it into core, so I definitely understand that.

**Comment by purdeaandrei** [comment]
@drashna
I have rebased the branch on top of latest develop.
I have also renamed *power* to *usb_power*. So for example the main .c file is now quantum/usb_power.c, and the main variable holding the usb power state is called "usb_power_state".
The word 'tracking' is not in the source code content, only in some commit messages, the branch name, and the PR description.

Let me know if any further changes are needed.
Note: I have not yet tested this since the rebase. I will test soon, and I will drop another comment here when that happens.

**Comment by drashna** [comment]
Honestly, i'm still not too happy with the naming, as it doesn't really reflect *what* it is. 

`usb_device_state_*` feels much more accurate of a name for this (and is similar to the LUFA and ChibiOS event names, to boot).

**Comment by purdeaandrei** [comment]
@drashna Renamed as you suggested.

File: quantum/usb_device_state.h [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    USB_DEVICE_STATE_NO_INIT    = 0,  // We're in this state before calling power_init()
+    USB_DEVICE_STATE_INIT       = 1,  // Can consume up to 100mA
+    USB_DEVICE_STATE_CONFIGURED = 2,  // Can consume up to what is specified in configuration descriptor, typically 500mA
+    USB_DEVICE_STATE_SUSPEND    = 3   // Can consume only suspend current
+};
+
+extern enum usb_device_state usb_device_state;
+
+void notify_usb_device_state_change_kb(enum power_state power_state);
+void notify_usb_device_state_change_user(enum power_state power_state);
```
Comment (drashna) [inline quantum/usb_device_state.h:38]:
```suggestion
void notify_usb_device_state_change_kb(enum usb_device_state power_state);
void notify_usb_device_state_change_user(enum usb_device_state power_state);
```

**Review by drashna — CHANGES_REQUESTED** [review]

**Review by drashna — APPROVED** [review]

**Comment by purdeaandrei** [comment]
Rebased (not cause this PR needed it, but because the dependent 12692 PR had a conflict)

**Comment by tzarc** [comment]
Can you explain the unused `configurationNumber` parameters?

**Comment by purdeaandrei** [comment]
@tzarc In general, the configuration number matters to how much power the USB device is allowed to consume, because each USB device could have multiple configurations, only one of which is selected at any point in time (or none if not configured).
Each configuration descriptor has its own ".MaxPowerConsumption", so you could have in theory one configuration that consumes 500mA, and one that consumes only 100mA. At the moment everything is using just one configuration in QMK, so that's why it's unused for now.

One possible usecase in the future would be, if a keyboard has a 500mA-using solenoid driver, it could have one configuration to select 500mA, and another one to select 100mA, and disable the solenoid driver automatically, or set it into a 100mA current-limited mode. So when you plug the keyboard into a normal USB port, or powered hub, you get full solenoid operation, but in an unpowered usb hub, the solenoid is automatically disabled, or in slower current-limited mode.

Or in case of keyboards wit high current-consumption backlights, it could throttle the backlight intensity when plugged into a 100mA-only port.

I'm not 100% sure if default HID operating system drivers are smart enough to select the alternative 100mA configuration, when they can't enable the 500mA one, or if they will just disable the USB device, so my specific example may not be a good one, but it's still worth keeping the interface, in case we will add multiple-configuration support in the future for other reasons.

**Review by jackhumbert — APPROVED** [review]

**Review by tzarc — APPROVED** [review]

**Pull request merged**